src/Pricing/PriceCalculationVisitor.php

```diff
@@ -93,292 +98,294 @@ public function visit(Delivery $delivery, PricingRuleSet $ruleSet): Output
+            $language = new ExpressionLanguage();
         }
+
+        $result = $rule->apply($expressionLanguageValues, $language);
+
+        $this->logger->info(sprintf('processProductOptionValue; result %d (rule "%s")', $result, $rule->getExpression()), [
+            'target' => $rule->getTarget(),
+        ]);
+
+        //For now; km and package-based rules will contain total in $price
+        // return price per km or package and quantity separately?
+        $productOptionValue->setPrice($result);
+
+        return new ProductOptionValueWithQuantity($productOptionValue, 1);
     }
 
     /**
-     * @param ProductVariant[] $taskProductVariants
+     * @param ProductVariantInterface[] $taskProductVariants
+     * @return ProductVariantInterface[]
      */
-    private function process(array $taskProductVariants, ?ProductVariant $deliveryProductVariant): Order
+    private function process(array $taskProductVariants, ?ProductVariantInterface $deliveryProductVariant): array
     {
-        $taskItems = [];
         $taskItemsTotal = 0;
 
         foreach ($taskProductVariants as $productVariant) {
             $this->processProductVariant($productVariant, 0);
 
-            $orderItem = new OrderItem($productVariant);
-
-            // Later on: group the same product variants into one OrderItem
-            $orderItem->setTotal($productVariant->getPrice());
-
-            $taskItems[] = $orderItem;
-            $taskItemsTotal += $orderItem->getTotal();
+            $taskItemsTotal += $productVariant->getOptionValuesPrice();
         }
 
-
-        $items = $taskItems;
-        $itemsTotal = $taskItemsTotal;
-
         if ($deliveryProductVariant) {
             $this->processProductVariant($deliveryProductVariant, $taskItemsTotal);
-
-            $orderItem = new OrderItem($deliveryProductVariant);
-            $orderItem->setTotal($deliveryProductVariant->getPrice());
-
-            $items[] = $orderItem;
-            $itemsTotal += $orderItem->getTotal();
         }
 
-        $order = new Order($items);
-        $order->setItemsTotal($itemsTotal);
-
-        return $order;
+        return array_merge($taskProductVariants, $deliveryProductVariant ? [$deliveryProductVariant] : []);
     }
 
-    private function processProductVariant(ProductVariant $productVariant, int $previousItemsTotal): void
+    private function processProductVariant(ProductVariantInterface $productVariant, int $previousItemsTotal): void
     {
         $subtotal = $previousItemsTotal;
 
-        foreach ($productVariant->getProductOptions() as $productOption) {
-            $priceAdditive = $productOption->getPriceAdditive();
-            $priceMultiplier = $productOption->getPriceMultiplier();
+        /**
+         * @var ProductOptionValueInterface $productOptionValue
+         */
+        foreach ($productVariant->getOptionValues() as $productOptionValue) {
+
+            if ('CPCCL-ODDLVR-PERCENTAGE' === $productOptionValue->getOptionCode()) {
```

Comment:
I didn't understand how this product option value is created?

Reply:
Each pricing rule is linked to a `product option value`: https://github.com/coopcycle/coopcycle-web/pull/5041/files#diff-6d0bf011cb02bc7af29f5165a12ba776544276b344a75a571000ef8c54c08551R76-R82

A `product option value` is in turn linked to a `product option`.

There are four `product option`s at the moment, one for each type of a price expression:  https://github.com/coopcycle/coopcycle-web/pull/5041/files#diff-2f4eb521f276a6cf3656e4edfc7da899063408d39c0b75f2b0b19fb3b68db0dfR42-R57

'CPCCL-ODDLVR-PERCENTAGE' is a `product option` for pricing rules with a price expression: `price_percentage(..`  https://github.com/coopcycle/coopcycle-web/pull/5041/files#diff-2f4eb521f276a6cf3656e4edfc7da899063408d39c0b75f2b0b19fb3b68db0dfR44-R46